### PR TITLE
Support Artifactory REST APIv2

### PR DESCRIPTION
**Pull request by thenewnano**
The V2 REST APIs apply for Artifactory version 6.6.0
and above, the user can pass the version explicitly.

Currently this only affects Permissions.

## Description
The permission scheme has changed in new version of artifactory, I had to find a way to behave based on API version supported by the server. 
Trying dynamic check of the server version I ended up in 5-6 server calls for each class that was called in class Artifactory.__init__
So At the end I opted for adding an argument that can be send to the constructor at the time the class is instantiated 

https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API+V2

I turned of the complexity fail on number of arguments, as all the ways around it would add more complexity. but I'm open to suggestions. 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How has it been tested ?

Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Against a live deployment
- [x] Unit tests updated and checked

## Checklist:

- [x] My PR is ready for prime time! Otherwise use the ["Draft PR" feature](https://help.github.com/en/articles/about-pull-requests#draft-pull-requests)
- [x] All commits have a correct title
- [x] Readme has been updated
- [x] Quality tests are green (see Codacy)
- [x] Automated tests are green (see pipeline)

### CODE REVIEW

**Comment by thenewnano** [comment]
@anancarv I would appreciate if you or another maintainer could review my PR, as I plan to put this lib in production this week it would be great if we could put together a release.  
I have put roughly 8 hours into this including the setup and test etc and like the level of quality you have set up, me and my team will probably contribute more to the project the coming weeks.

I would really want to avoid forking and publishing, my version though, which I'll have to do in a few days.

**Comment by anancarv** [comment]
Hi @thenewnano , thanks for your contribution. We'll review your PR as soon as possible.
I understand that you have production needs, we'll do our best to give you a feedback soon

File: pyartifactory/objects.py
Comment (ee987) [inline pyartifactory/objects.py:182]: Any reason to have both create and update? Should they be included into one, with upsert-argument instead?

File: pyartifactory/objects.py
Comment (ee987) [inline pyartifactory/objects.py:218]: Not sure but i think 400 doesn't necessarily mean "does not exist".

File: tests/test_permissions.py
Comment (anancarv) [inline tests/test_permissions.py:104]: I think you can get rid of the `mocker` parameter here since it is not used.

**Review by anancarv — CHANGES_REQUESTED** [review]
Very nice work. I only have a few remarks but once they've been fixed it's all good for me. Thanks again for this PR ;)

File: pyartifactory/objects.py
Comment (anancarv) [reply pyartifactory/objects.py:182]: In this library, we follow the Artifactory API endpoints that are `CREATE, GET, UPDATE and DELETE` options. We do not want to disturb the users by including create and update into one call.
https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API+V2#ArtifactoryRESTAPIV2-CreatePermissionTarget

File: pyartifactory/objects.py
Comment (anancarv) [reply pyartifactory/objects.py:218]: Normally it doesn't, but the Artifactory API sometimes sends a 400 error code. That's why we added it.

File: pyartifactory/objects.py
Comment (anancarv) [reply pyartifactory/objects.py:246]: I understand, but we made the choice of raising a PermissionNotFoundException when the permission is missing. If we change this behavior here, we would would change it for all delete operations. It would need a new PR especially for this. You're welcome to propose it and I'll be glad to review it.

**Comment by anancarv** [comment]
I've seen it afterwards, but you have a little typo in the README:
```
when you instantiate the calss.
```

**Review by anancarv — DISMISSED** [review]
Looks good to me !! Let's just wait for a second point of view from @nymous before it's merged.

File: pyartifactory/objects.py
Comment (nymous) [inline pyartifactory/objects.py:205]: Now that we use aliases in our models we need to update the way we deserialize it:
```suggestion
            self._put(f"api/{self._uri}/{permission_name}", json=permission.dict(by_alias=True))
```

File: pyartifactory/objects.py
Comment (nymous) [inline pyartifactory/objects.py:255]: Same thing here
```suggestion
        self._put(f"api/{self._uri}/{permission_name}", json=permission.dict(by_alias=True))
```

File: tests/test_permissions.py
Comment (nymous) [inline tests/test_permissions.py:41]: Now that we know that the JSON changed, we need to update it in the tests too!
```suggestion
PERMISSIONV2 = PermissionV2(
    **{
        "name": "test_permission",
        "repo": {
            "repositories": ["test_repository"],
            "actions": {
                "users": {"test_user": ["read", "annotate", "write", "delete",]},
                "groups": {"developers": ["read", "annotate", "write", "delete",],},
            },
        },
        "include-patterns": ["**"],
        "exclude-patterns": [],
    }
)
```

**Review by nymous — CHANGES_REQUESTED** [review]
Hello @thenewnano, thank you for your contribution!

I also reviewed the code, and you might see that it has kind of 2 "parts" ^^
First I fixed many little typos, missing imports in the examples... so just suggested changes without any comment (*). There is still an important issue in the tests (in that they don't test much because you compare the permission object to itself :sweat_smile:).
Then I tested the code on our Jfrog Cloud instance, and realized that there was a big (but subtle!) change in Artifactory API, which meant the code was not working as intended! I went into more details in the review below.

Sorry for the many comments :grimacing:

Now for a more general overview of this new feature: you implemented the "repo" part of the v2 permission API, I think we should also quickly implement the "build" part, because if you update an existing permission which has both "repo" and "build" but don't set one or the other, the corresponding sub-permission is removed.
We can however merge this PR and do another one for the "build", if you want (maybe not release before the feature is complete, @anancarv?)


(*): This is a question I ask myself for this kind of remarks (typo and minor docs changes): should I add a PR comment for each one, or push a commit fixing everything because it will bother you less  ^^ (but you might feel attacked if I commit on your branch, and I could break something if you are also working on it but didn't push yet). This is a philosophical question, asking you how you would feel ^^ (and I know it would be different for each contributor)

**Comment by sciyako** [comment]
@nymous Thank you for the feedback, yes we want builds as well implemented.

**Comment by thenewnano** [comment]
@nymous it should be all good now! Thank you for the great review and the suggestions.

**Review by nymous — APPROVED** [review]
Great! Will you work on the "build" support right after this PR, or should we do it?

**Comment by anancarv** [comment]
@thenewnano @sciyako , as said before by @nymous the publication of the `PermissionV2` feature may cause issues with the `builds`. Indeed, if you update an existing permission which has both "repo" and "build" but don't set one or the other, the corresponding sub-permission is removed.
Do you want us to publish it as it is and then later work on the `builds` part ? Or, do you want us to first implement the `builds` before publishing the new version of the library. We're asking because you seemed to want to feature ASAP.

**Comment by thenewnano** [comment]
Thank you @nymous  and @anancarv this has been fun and inspiring first contribution. 
Currently I don't have permission to manage builds on the new server I am intending to develop my code against. 

So if you can start with a branch that I can help with it would be nice too, if you have not been able to start before I have access to a server to test the build implementation with then I might start. 

It would be cool if you do a release to pypi so I can start using this module instead of the one we use today at least where the feature are already available. 
There are some other features that I might add needed in the project we do now,  features needed there will have priority for my contribution  to add here instead of the internal closed source module.

**Comment by anancarv** [comment]
Ok then, a new release has been published `v1.9.0`. You can now use the `permissions v2` feature.

**Pull request merged**
